aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -837,40 +1136,174 @@ void flash_attention_kernel_impl(
-    if (!attn_mask.has_value()) {
-      if (q_seq_len >= 768) {
-        cpu_flash_attention<scalar_t, scalar_t, 256, 512>(
-          output, logsumexp, query, key, value,
-          dropout_p, is_causal, attn_mask, scale);
-      } else if (q_seq_len >= 192) {
-        cpu_flash_attention<scalar_t, scalar_t, 64, 512>(
-          output, logsumexp, query, key, value,
-          dropout_p, is_causal, attn_mask, scale);
-      } else {
-        cpu_flash_attention<scalar_t, scalar_t, 32, 512>(
-          output, logsumexp, query, key, value,
-          dropout_p, is_causal, attn_mask, scale);
-      }
-    } else {
-      AT_DISPATCH_MASK_TYPES(attn_mask.value().scalar_type(), "flash_attention_mask", [&]() {
-        if (q_seq_len >= 768) {
-          cpu_flash_attention<scalar_t, mask_t, 256, 512>(
-            output, logsumexp, query, key, value,
-            dropout_p, is_causal, attn_mask, scale);
-        } else if (q_seq_len >= 192) {
-          cpu_flash_attention<scalar_t, mask_t, 64, 512>(
-            output, logsumexp, query, key, value,
-            dropout_p, is_causal, attn_mask, scale);
-        } else {
-          cpu_flash_attention<scalar_t, mask_t, 32, 512>(
-            output, logsumexp, query, key, value,
-            dropout_p, is_causal, attn_mask, scale);
-        }
-      });
-    }
-  });
+  // When q_seq_len and k_seq_len are long enough,
+  // cpu_flash_attention_with_pack has better performance.
+  if ((query.scalar_type() == kBFloat16 &&
+       at::native::cpublas::need_pack(kBFloat16) && q_seq_len >= 320 &&
+       k_seq_len >= 320 && nhead >= 4 && headSize >= 128) ||
+      (query.scalar_type() == kHalf && at::native::cpublas::need_pack(kHalf) &&
+       q_seq_len >= 64 && k_seq_len >= 64 && headSize >= 64)) {
+    AT_DISPATCH_REDUCED_FLOATING_TYPES(
+        query.scalar_type(), "flash_attention", [&] {
+          if (!attn_mask.has_value()) {
+            if (q_seq_len >= 768) {
+              cpu_flash_attention<scalar_t, scalar_t, 256, 512, true>(
+                  output,
+                  logsumexp,
+                  query,
+                  key,
+                  value,
+                  dropout_p,
+                  is_causal,
+                  attn_mask,
+                  scale);
+            } else if (q_seq_len >= 192) {
+              cpu_flash_attention<scalar_t, scalar_t, 64, 512, true>(
+                  output,
+                  logsumexp,
+                  query,
+                  key,
+                  value,
+                  dropout_p,
+                  is_causal,
+                  attn_mask,
+                  scale);
+            } else {
+              cpu_flash_attention<scalar_t, scalar_t, 32, 512, true>(
+                  output,
+                  logsumexp,
+                  query,
+                  key,
+                  value,
+                  dropout_p,
+                  is_causal,
+                  attn_mask,
+                  scale);
+            }
```

Comment:
These added specialized calls are very similar to the original ones. Is it possible to factor out some function to avoid duplication?

Reply:
Use a macro `FLASH_ATTENTION_KERNEL` to avoid duplication.